content/blog/2022/async-2.md

```diff
@@ -0,0 +1,191 @@
+prior to including pmem2.
+
+```c
+#define PMEM2_USE_MINIASYNC 1
+#include "libpmem2.h"
+```
+
+This is because we've made a conscious decision to make this software feature
+entirely optional, at least for now. We also made sure that the core miniasync
+abstractions (futures, vdm) are header-only. This will make it easier
+for existing users to upgrade PMDK version if they don't want miniasync.
+It will also make it possible for us to package pmem2 into various linux and
+windows repositories without having first to package miniasync.
+
+## Future-looking plans and ideas
+
+As I've said above, our work on asynchronous interfaces doesn't stop with pmem2.
+We intend on leveraging miniasync throughout the stack of PMDK libraries. The
+first example of that is in pmemstream, our work-in-progress solution for
+efficient log storage and processing.
+
+In pmemstream, we want to make a general abstraction for all types of logs. We
+also want to make sure that our library doesn't just run on PMem, but also
+supports a wider variety of heterogeneous memory and storage systems.
+To accomplish that, we are designing our on-media layout in a way that seamlessly
+supports [buffered durable linearizability][linearizability]. This means
+that we can make writes fully asynchronous while still making sure that everything
+remains consistent, even in the presence of failures. And obviously we are going to
+use miniasync to create that asynchronous API.
+
+Our other plans include using asynchronous APIs in libpmemobj, our transactional
+object storage system. One of the enhancements that we are thinking about is
+improving our undo and redo logging implementations to allow concurrent processing
+of log entries, potentially reducing the time it takes for transactions to be
+committed or aborted.
+
+And, of course, we also want to use miniasync in our user-space block storage
+API - libpmemblk.
+
+## Asynchronous block storage APIs
+
+The pmemblk library is a failure-atomic implementation of a block storage
+interface on top of Persistent Memory. This is needed because PMem, unlike
+standard enterprise disks, does not provide sector atomicity (because there
+are no sectors). If you were to simply create a huge array of blocks and
+write to it directly, you might have data consistency issues when the system crashes.
+This behavior might be fine for software designed to tolerate failures,
+but not for anything else. Our library addresses this problem through the use
+of [Block Translation Table][btt] algorithm.
+
+However, as I've described in my [first article][async-blog], using PMem as block
+storage has an additional cost of using the CPU to perform the actual
+I/O. This also applies to libpmemblk. Thankfully, with DSA and miniasync, we can
+address this problem.
+
+![overview](/images/posts/async-2-pmemblk.png)
+
+Logically, a write in the BTT algorithm I mentioned earlier is composed
+of four steps:
+
+1. Allocating a new block for the new write.
```

Comment:
I think, it could be changed to `Allocate` since the rest of the verbs from the list are in the from without ing.